### PR TITLE
Add RemoteConfig

### PR DESCRIPTION
**Pull request by marinacoelho**
This pull request adds RemoteConfig to project, creates ConfigurationService to fetch data and ConfigurationDefaults to store default key and value pairs. With the new service, it's possible to fetch ShowDetailsButton config from Firebase and reflect it in the todo lists view.

### CODE REVIEW

File: code/frontend/MakeItSo/MakeItSo/Configuration/Services/ConfigurationService.swift [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        RemoteConfig.remoteConfig().configSettings = settings
+    }
+    
+    func fetchConfigurationData() {
+        RemoteConfig.remoteConfig().fetch { (status, error) -> Void in
+          if status == .success {
+              RemoteConfig.remoteConfig().activate()
+              self.showDetailsButton = RemoteConfig.remoteConfig().configValue(forKey: ConfigurationDefaults.showDetailsButtonKey).boolValue
+          } else {
+              self.logger.debug("Could not fetch configuration data")
```
Comment (peterfriese) [inline code/frontend/MakeItSo/MakeItSo/Configuration/Services/ConfigurationService.swift:44]: Might not need to prefix this with self.

**Review by peterfriese — CHANGES_REQUESTED** [review]
Looking good, just a few nits.

One additional idea: we could add a second parameter for the shape of the checkbox in the list (round vs square) so people have another thing to play around with. This might also be a good start for A/B testing (will people finish more tasks if the check box is round?)

File: code/frontend/MakeItSo/MakeItSo/Configuration/Services/ConfigurationService.swift [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+import os
+import FirebaseRemoteConfig
+
+public class ConfigurationService: ObservableObject {
+    private let logger = Logger(subsystem: "com.google.firebase.workshop.MakeItSo", category: "configuration")
+    
+    @Published var showDetailsButton: Bool = ConfigurationDefaults.showDetailsButtonValue
+
+    init() {
+        RemoteConfig.remoteConfig().setDefaults(fromPlist: "remote_config_defaults")
```
[the remark this answers is not in the document]
Comment (marinacoelho) [reply code/frontend/MakeItSo/MakeItSo/Configuration/Services/ConfigurationService.swift:30]: I just used the original name of the file, but I can rename it!

File: code/frontend/MakeItSo/MakeItSo/Configuration/Services/ConfigurationService.swift [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        
+        // Only used for testing, should never deploy this code to production:
```
[the remark this answers is not in the document]
Comment (marinacoelho) [reply code/frontend/MakeItSo/MakeItSo/Configuration/Services/ConfigurationService.swift:32]: Yes, I'll do it.

**Comment by marinacoelho** [comment]
> 

Good idea, I added a "todoCheckShape" config to this PR.

**Comment by marinacoelho** [comment]
> Looking good, just a few nits.
> 
> One additional idea: we could add a second parameter for the shape of the checkbox in the list (round vs square) so people have another thing to play around with. This might also be a good start for A/B testing (will people finish more tasks if the check box is round?)

Good idea, I added a "todoCheckShape" config to this PR.

File: code/frontend/MakeItSo/MakeItSo/Configuration/Services/ConfigurationDefaults.swift [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+
+struct ConfigurationDefaults {
+    static let showDetailsButtonKey = "ShowDetailsButton"
```
[the remark this answers is not in the document]
Comment (marinacoelho) [reply code/frontend/MakeItSo/MakeItSo/Configuration/Services/ConfigurationDefaults.swift]: Done!

File: code/frontend/MakeItSo/MakeItSo/Configuration/Services/ConfigurationService.swift [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+import os
+import FirebaseRemoteConfig
+
+public class ConfigurationService: ObservableObject {
+    private let logger = Logger(subsystem: "com.google.firebase.workshop.MakeItSo", category: "configuration")
```
[the remark this answers is not in the document]
Comment (marinacoelho) [reply code/frontend/MakeItSo/MakeItSo/Configuration/Services/ConfigurationService.swift]: Done!

File: code/frontend/MakeItSo/MakeItSo/Configuration/Services/ConfigurationService.swift [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        settings.minimumFetchInterval = 0
+        RemoteConfig.remoteConfig().configSettings = settings
+    }
+    
+    func fetchConfigurationData() {
+        RemoteConfig.remoteConfig().fetch { (status, error) -> Void in
+          if status == .success {
+              RemoteConfig.remoteConfig().activate()
+              self.showDetailsButton = RemoteConfig.remoteConfig().configValue(forKey: ConfigurationDefaults.showDetailsButtonKey).boolValue
+          } else {
```
[the remark this answers is not in the document]
Comment (marinacoelho) [reply code/frontend/MakeItSo/MakeItSo/Configuration/Services/ConfigurationService.swift]: Done!

File: code/frontend/MakeItSo/MakeItSo/Configuration/Services/ConfigurationService.swift (hunk 1 above)
Comment (marinacoelho) [reply code/frontend/MakeItSo/MakeItSo/Configuration/Services/ConfigurationService.swift]: I tried to remove the prefix, but apparently it's needed :/

File: code/frontend/MakeItSo/MakeItSo/Configuration/Services/ConfigurationService.swift [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    #endif
+  }
+    
+  func fetchConfigurationData() async throws {
+    let status = try await RemoteConfig.remoteConfig().fetch()
+    if status == .success {
+      try await RemoteConfig.remoteConfig().activate()
+      DispatchQueue.main.async {
+        self.showDetailsButton = RemoteConfig.remoteConfig().configValue(forKey: ConfigurationDefaults.showDetailsButtonKey).boolValue
+        self.todoCheckShape = RemoteConfig.remoteConfig().configValue(forKey: ConfigurationDefaults.todoCheckShapeKey).stringValue ?? ConfigurationDefaults.todoCheckShapeValue
```
Comment (peterfriese) [inline code/frontend/MakeItSo/MakeItSo/Configuration/Services/ConfigurationService.swift:46]: We also support Codable now for RC, so you can do this:

```
self.todoCheckShape = try RemoteConfig.remoteConfig()[ConfigurationDefaults.todoCheckShapeKey].decoded(asType: TodoCheckShape.self)
```

In `ConfigurationDefaults.swift`, you need to define this:

```
enum TodoCheckShape: String, Decodable {
  case circle
  case square
  
  func iconName(completed: Bool) -> String {
    let shape = self.rawValue
    return completed ? "checkmark.\(shape).fill" : shape
  }
}
```

And then , in `TodoListRowView.swift`, you can do this:

```
Image(systemName: viewModel.todoCheckShape.iconName(completed: viewModel.todo.completed))
```

This makes the whole thing more type safe.

This also requires some minor changes to the view model and service:

```
class TodoListRowViewModel: ObservableObject {
  @Published var todoCheckShape = TodoCheckShape.circle
    
  @LazyInjected private var repository: TodosRepository
  @LazyInjected private var configurationService: ConfigurationService
  
  @Binding var todo: Todo
  
  init(todo: Binding<Todo>) {
    self._todo = todo
      
    configurationService
      .$todoCheckShape
      .assign(to: &$todoCheckShape)
  }
  
  func toggleCompletion() {
    var userTodo = todo
    userTodo.completed.toggle()
    repository.updateTodo(userTodo)
  }
  
}
```

```
// ...
import FirebaseRemoteConfigSwift

public class ConfigurationService: ObservableObject {
  private let logger = Logger(subsystem: "com.google.firebase.workshop.MakeItSo", category: "configuration")
    
  @Published var showDetailsButton: Bool = ConfigurationDefaults.showDetailsButtonValue
  @Published var todoCheckShape: TodoCheckShape = ConfigurationDefaults.todoCheckShapeValue
  // ...
```

**Review by peterfriese — CHANGES_REQUESTED** [review]
Looking good! I think you might need to rebase this first to resolve the conflicts (and there are a couple of things I changed that you probably are interested in).

Other than that, I made a few suggestions around the handling of exceptions and async/await, and how to use Codable (which makes the most sense for the check box shape, as this gives us nice type safety). For the feature flag, we can continue using the `Bool` type.

File: code/frontend/MakeItSo/MakeItSo.xcodeproj/project.pbxproj [hunk 8]
```diff
@@ -3,600 +3,411 @@
...
-		88FEECFA27280F3D00ED368C /* ReminderListRowView.swift in Sources */ = {isa = PBXBuildFile; fileRef = 88FEECF927280F3D00ED368C /* ReminderListRowView.swift */; };
-		88FEECFB27280F3D00ED368C /* ReminderListRowView.swift in Sources */ = {isa = PBXBuildFile; fileRef = 88FEECF927280F3D00ED368C /* ReminderListRowView.swift */; };
+		1223C1FF28AFE168006037A2 /* ConfigurationService.swift in Sources */ = {isa = PBXBuildFile; fileRef = 1223C1FE28AFE168006037A2 /* ConfigurationService.swift */; };
+		1223C20228AFE70F006037A2 /* FirebaseRemoteConfig in Frameworks */ = {isa = PBXBuildFile; productRef = 1223C20128AFE70F006037A2 /* FirebaseRemoteConfig */; };
+		1223C20428AFE70F006037A2 /* FirebaseRemoteConfigSwift in Frameworks */ = {isa = PBXBuildFile; productRef = 1223C20328AFE70F006037A2 /* FirebaseRemoteConfigSwift */; };
+		1223C20828AFF7A7006037A2 /* ConfigurationDefaults.swift in Sources */ = {isa = PBXBuildFile; fileRef = 1223C20728AFF7A7006037A2 /* ConfigurationDefaults.swift */; };
+		125A1A1C28BF7B3800E8A538 /* RemoteConfigDefaults.plist in Resources */ = {isa = PBXBuildFile; fileRef = 125A1A1B28BF7B3800E8A538 /* RemoteConfigDefaults.plist */; };
```
[the remark this answers is not in the document]
Comment (marinacoelho) [reply code/frontend/MakeItSo/MakeItSo.xcodeproj/project.pbxproj:14]: Ok, I'll do it!

File: code/frontend/MakeItSo/MakeItSo/Configuration/Services/ConfigurationService.swift [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    RemoteConfig.remoteConfig().setDefaults(fromPlist: "RemoteConfigDefaults")
+      
+    #if DEBUG
+      let settings = RemoteConfigSettings()
+      settings.minimumFetchInterval = 0
+      RemoteConfig.remoteConfig().configSettings = settings
+    #endif
+  }
+    
+  func fetchConfigurationData() async throws {
```
[the remark this answers is not in the document]
Comment (marinacoelho) [reply code/frontend/MakeItSo/MakeItSo/Configuration/Services/ConfigurationService.swift:40]: Great, changing it!

File: code/frontend/MakeItSo/MakeItSo/App/MakeItSoApp.swift [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    return true
+  }
+}
+
+@main
+struct MakeItSoApp: App {
+  @UIApplicationDelegateAdaptor(AppDelegate.self) var delegate
+  @LazyInjected var authenticationService: AuthenticationService
```
[the remark this answers is not in the document]
Comment (marinacoelho) [reply code/frontend/MakeItSo/MakeItSo/App/MakeItSoApp.swift]: Removed.

File: code/frontend/MakeItSo/MakeItSo/Configuration/Services/ConfigurationService.swift [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+public class ConfigurationService: ObservableObject {
+  private let logger = Logger(subsystem: "com.google.firebase.workshop.MakeItSo", category: "configuration")
+    
+  @Published var showDetailsButton: Bool = ConfigurationDefaults.showDetailsButtonValue
+  @Published var todoCheckShape: String = ConfigurationDefaults.todoCheckShapeValue
+
+  init() {
+    RemoteConfig.remoteConfig().setDefaults(fromPlist: "RemoteConfigDefaults")
+      
+    #if DEBUG
```
[the remark this answers is not in the document]
Comment (marinacoelho) [reply code/frontend/MakeItSo/MakeItSo/Configuration/Services/ConfigurationService.swift]: Done!

File: code/frontend/MakeItSo/MakeItSo/Configuration/Services/ConfigurationService.swift [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+import FirebaseRemoteConfig
+
+public class ConfigurationService: ObservableObject {
+  private let logger = Logger(subsystem: "com.google.firebase.workshop.MakeItSo", category: "configuration")
+    
+  @Published var showDetailsButton: Bool = ConfigurationDefaults.showDetailsButtonValue
+  @Published var todoCheckShape: String = ConfigurationDefaults.todoCheckShapeValue
+
+  init() {
+    RemoteConfig.remoteConfig().setDefaults(fromPlist: "RemoteConfigDefaults")
```
[the remark this answers is not in the document]
Comment (marinacoelho) [reply code/frontend/MakeItSo/MakeItSo/Configuration/Services/ConfigurationService.swift]: Great idea, I moved it to the dependency injection file.

**Comment by peterfriese** [comment]
LGTM, merging.

**Pull request merged**